rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/action/TbDeviceStateNode.java

```diff
@@ -0,0 +1,187 @@
+        ctx.getClusterService().pushMsgToCore(
+                ctx.getTenantId(), msg.getOriginator(), toCoreMsg, getMsgProcessedCallback(ctx, msg)
+        );
+    }
+
+    private void sendDeviceActivityMsg(TbContext ctx, TbMsg msg) {
+        var tenantUuid = ctx.getTenantId().getId();
+        var deviceUuid = msg.getOriginator().getId();
+        var deviceActivityMsg = TransportProtos.DeviceActivityProto.newBuilder()
+                .setTenantIdMSB(tenantUuid.getMostSignificantBits())
+                .setTenantIdLSB(tenantUuid.getLeastSignificantBits())
+                .setDeviceIdMSB(deviceUuid.getMostSignificantBits())
+                .setDeviceIdLSB(deviceUuid.getLeastSignificantBits())
+                .setLastActivityTime(System.currentTimeMillis())
+                .build();
+        var toCoreMsg = TransportProtos.ToCoreMsg.newBuilder()
+                .setDeviceActivityMsg(deviceActivityMsg)
+                .build();
+        ctx.getClusterService().pushMsgToCore(
+                ctx.getTenantId(), msg.getOriginator(), toCoreMsg, getMsgProcessedCallback(ctx, msg)
+        );
+    }
+
+    private void sendDeviceDisconnectMsg(TbContext ctx, TbMsg msg) {
+        var tenantUuid = ctx.getTenantId().getId();
+        var deviceUuid = msg.getOriginator().getId();
+        var deviceDisconnectMsg = TransportProtos.DeviceDisconnectProto.newBuilder()
+                .setTenantIdMSB(tenantUuid.getMostSignificantBits())
+                .setTenantIdLSB(tenantUuid.getLeastSignificantBits())
+                .setDeviceIdMSB(deviceUuid.getMostSignificantBits())
+                .setDeviceIdLSB(deviceUuid.getLeastSignificantBits())
+                .build();
+        var toCoreMsg = TransportProtos.ToCoreMsg.newBuilder()
+                .setDeviceDisconnectMsg(deviceDisconnectMsg)
+                .build();
+        ctx.getClusterService().pushMsgToCore(
+                ctx.getTenantId(), msg.getOriginator(), toCoreMsg, getMsgProcessedCallback(ctx, msg)
+        );
+    }
+
+    private void sendDeviceInactivityMsg(TbContext ctx, TbMsg msg) {
+        var tenantUuid = ctx.getTenantId().getId();
+        var deviceUuid = msg.getOriginator().getId();
+        var deviceInactivityMsg = TransportProtos.DeviceInactivityProto.newBuilder()
+                .setTenantIdMSB(tenantUuid.getMostSignificantBits())
+                .setTenantIdLSB(tenantUuid.getLeastSignificantBits())
+                .setDeviceIdMSB(deviceUuid.getMostSignificantBits())
+                .setDeviceIdLSB(deviceUuid.getLeastSignificantBits())
+                .build();
+        var toCoreMsg = TransportProtos.ToCoreMsg.newBuilder()
+                .setDeviceInactivityMsg(deviceInactivityMsg)
+                .build();
+        ctx.getClusterService().pushMsgToCore(
+                ctx.getTenantId(), msg.getOriginator(), toCoreMsg, getMsgProcessedCallback(ctx, msg)
+        );
+    }
+
+    private TbQueueCallback getMsgProcessedCallback(TbContext ctx, TbMsg msg) {
```

Comment:
I found 27 implementations of this callback.
Do we need another one?
If so, please create a new inner or outer class extending TbQueueCallback.
If you analyze the heap dump, you can count and filter exactly your object types. 
![image](https://github.com/thingsboard/thingsboard/assets/79898499/9f3bea2f-fcd0-4363-b50d-974975a98c28)

Reply:
I made one implementation public and used it in my code. Also, it required small changes in place that it was used before. Please see `SimpleTbQueueCallback` and `DefaultTbContext` classes